### ISSUE TITLE
`non_zero_suggestions` is triggered when converting from a nonzero value to a larger unsigned value

### ISSUE THREAD

**Issue by zacknewman** [rust-lang/rust-clippy#13757]
### Summary

`T::from` is getting confused with `G::from` when `T` is an unsigned integer (e.g., `u64`) and `G` is the nonzero version (e.g., `NonZeroU64`). For example converting a `NonZeroU32` into a `u64` triggers the lint incorrectly.

### Lint Name

non_zero_suggestions

### Reproducer

I tried this code:

```rust
#![deny(clippy::non_zero_suggestions)]
use core::num::NonZeroU32;
fn main() {
    _ = u64::from(NonZeroU32::new(10).unwrap().get());
}
```

I saw this happen:

```
error: consider using `NonZeroU64::from()` for more efficient and type-safe conversion
 --> src/main.rs:4:9
  |
4 |     _ = u64::from(NonZeroU32::new(10).unwrap().get());
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `NonZeroU64::from(NonZeroU32::new(10).unwrap())`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#non_zero_suggestions
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(clippy::non_zero_suggestions)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `bug` (bin "bug") due to 1 previous error
```

I expected to see this happen: Compilation to succeed since `u64::from` is not the same as `NonZeroU64::from`.

### Version

```text
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: x86_64-unknown-linux-gnu
release: 1.83.0
LLVM version: 19.1.1
```

**Comment by DylanBulfin** [comment]
I'm not an expert but this seems like expected behavior to me. What you're doing is creating a much more complicated, checked numeric type, with the overhead that entails, only to immediately convert it to a less-safe native type.

If you do need to do something like this you could instead do either of the below:
```
let _a: u64 = NonZeroU64::new(10).unwrap().get();
let _b: u64 = NonZeroU64::from(NonZeroU32::new(10).unwrap()).get();
```

**Comment by DylanBulfin** [comment]
I see what you mean now, looks like this was noticed during implementation but wasn't seen as worth fixing at the time: https://github.com/rust-lang/rust-clippy/pull/13167#discussion_r1738447824 

I don't think either of the approaches I suggested are better, I just mentioned them since they got around the basic example for me.

As for the suggestion being wrong, I think it's allowed in general for suggestions to be incorrect. There's a concept of 'applicability' of suggestions, and this suggestion is marked as `MaybeIncorrect`. There was some discussion around that in the PR as well https://github.com/rust-lang/rust-clippy/pull/13167#discussion_r1694722795. It seems like the applicability is only used when determining whether suggestions are auto-applied, not sure why that is.

I also see it's in the Restriction lint group so that might have something to do with it being overly limiting/incorrect.

Leaving this in case any of the context is helpful but I'm very new to Clippy development so I think this is over my head.

**Comment by nazar-pc** [comment]
I have a similar issue when converting `NonZeroU16` to `usize` for use with iterators. Suggestion clearly makes no sense since `.take()` on iterator expects `usize`:
```rust
   --> *.rs
    |
364 |     .take(usize::from(self.intermediate_shards.get()))
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `NonZeroUsize::from(self.intermediate_shards)`
```

**Comment by ronnodas** [comment]
I understand that this is a restriction lint, but it is very unexpected for it to trigger when the output is being used for indexing, just like with the `take()` example above. 
```rust
vector[usize::from(n.get())]
```
